app/src/main/java/org/kiwix/kiwixmobile/local_file_transfer/WifiDirectManager.kt

```diff
@@ -0,0 +1,395 @@
+    if (isFileSender && !hasSenderStartedConnection) {
+      this.senderSelectedPeerDevice = senderSelectedPeerDevice
+      alertDialogShower.show(
+        FileTransferConfirmation(senderSelectedPeerDevice.deviceName), {
+          hasSenderStartedConnection = true
+          connect()
+          displayToast(R.string.performing_handshake, Toast.LENGTH_LONG)
+        })
+    }
+  }
+
+  private fun connect() {
+    val config = WifiP2pConfig().apply {
+      deviceAddress = senderSelectedPeerDevice.deviceAddress
+      wps.setup = WpsInfo.PBC
+    }
+    manager.connect(channel, config, object : ActionListener {
+      override fun onSuccess() {
+        // UI updated from broadcast receiver
+      }
+
+      override fun onFailure(reason: Int) {
+        val errorMessage = getErrorMessage(reason)
+        Log.d(TAG, activity.getString(R.string.connection_failed) + ": " + errorMessage)
+        displayToast(R.string.connection_failed, Toast.LENGTH_LONG)
+      }
+    })
+  }
+
+  private fun performHandshakeWithSelectedPeerDevice() {
+    if (BuildConfig.DEBUG) {
+      Log.d(TAG, "Starting handshake")
+    }
+    peerGroupHandshakeAsyncTask = PeerGroupHandshakeAsyncTask(this)
+      .also { it.execute() }
+  }
+
+  val totalFilesForTransfer: Int
+    get() = filesForTransfer.size
+
+  fun getFilesForTransfer() = filesForTransfer
+
+  fun setFilesForTransfer(fileItems: ArrayList<FileItem>) {
+    filesForTransfer = fileItems
+  }
+
+  val zimStorageRootPath
+    get() = sharedPreferenceUtil.prefStorage + "/Kiwix/"
+
+  fun getFileReceiverDeviceAddress() = fileReceiverDeviceAddress
+
+  fun setClientAddress(clientAddress: InetAddress) {
+
+    // If control reaches here, means handshake was successful
+    selectedPeerDeviceInetAddress = clientAddress
+    startFileTransfer()
+  }
+
+  private fun startFileTransfer() {
+    if (isGroupFormed) {
+      if (isFileSender) {
+        Log.d(LocalFileTransferActivity.TAG, "Starting file transfer")
+        fileReceiverDeviceAddress =
+          if (isGroupOwner) selectedPeerDeviceInetAddress else groupOwnerAddress
+        displayToast(R.string.preparing_files, Toast.LENGTH_LONG)
+        senderDeviceAsyncTask = SenderDeviceAsyncTask(this, activity).also {
+          it.execute()
+        }
+      } else {
+        callbacks.onFilesForTransferAvailable(filesForTransfer)
+        receiverDeviceAsyncTask = ReceiverDeviceAsyncTask(this).also {
+          it.execute()
+        }
+      }
+    }
+  }
+
+  fun changeStatus(itemIndex: Int, @FileStatus status: Int) {
+    filesForTransfer[itemIndex].fileStatus = status
+    callbacks.onFileStatusChanged(itemIndex)
+    if (status == FileStatus.ERROR) {
+      displayToast(
+        R.string.error_transferring, filesForTransfer[itemIndex].fileName,
+        Toast.LENGTH_SHORT
+      )
+    }
+  }
+
+  private fun cancelAsyncTasks(vararg tasks: AsyncTask<*, *, *>?) =
+    tasks.all {
```

Comment:
you don't want to be using `all` that is for checking all of the items in a list a condition is true. Seeing how we don't care about the return value it is fine to just use a `forEach`